src/web/lib/messagePicker.ts

```diff
@@ -0,0 +1,123 @@
+import { record } from '@root/src/web/browser/ophan/ophan';
+
+export type MaybeFC = React.FC | null;
+type ShowMessage = (meta?: any) => MaybeFC;
+
+export type CanShowResult = {
+	result: boolean;
+	meta?: any;
+};
+
+export type Candidate = {
+	id: string;
+	show: ShowMessage;
+	canShow: () => Promise<CanShowResult>;
+};
+
+export type CandidateConfig = {
+	candidate: Candidate;
+	timeoutMillis: number | null;
+};
+
+type CandidateConfigWithTimeout = CandidateConfig & {
+	cancelTimeout: () => void;
+};
+
+export type SlotConfig = {
+	candidates: CandidateConfig[];
+	name: string;
+};
+
+const recordMessageTimeoutInOphan = (messageId: string, slotName: string) =>
+	record({
+		component: `${slotName}-picker-timeout-dcr`,
+		value: messageId,
+	});
+
+const timeoutify = (
+	candidateConfig: CandidateConfig,
+	slotName: string,
+): CandidateConfigWithTimeout => {
+	let timer: number | undefined;
+
+	const canShow = (): Promise<CanShowResult> =>
+		new Promise((resolve) => {
+			if (candidateConfig.timeoutMillis) {
+				timer = window.setTimeout(() => {
+					recordMessageTimeoutInOphan(
+						candidateConfig.candidate.id,
+						slotName,
+					);
+					resolve({ result: false });
+				}, candidateConfig.timeoutMillis);
+			}
+
+			candidateConfig.candidate
+				.canShow()
+				.then((result) => resolve(result));
+		});
+
+	const cancelTimeout = () => timer && clearTimeout(timer);
```

Comment:
Here too, I'm a little confused (sorry!)

This cancels the timeouts, but not the actual logic. So all the pending candidates will continue executing as normal. But since the cancel handle is called when a message is selected from the available candidates, should this not also resolve the can show promise with a failure?

Reading ahead, I think my issue is that I followed this from the "clearAllTimeouts" invocation. This makes sense from a "cancel the timeout if `canShow` succeeded" perspective. If *this* is the goal though, why not cancel the timeout in the `then` on line 57? Why are we depending on our caller to tidy up this piece of internal state for us?